File: rainbow.py
from OpenGL.GL import *
from OpenGL.GLUT import *
from OpenGL.GLU import *
import math
import time

# Window dimensions
width, height = 800, 600

# Rainbow state
current_band = 0

# Rainbow colors (VIBGYOR - from inner to outer)
rainbow_colors = [
    (0.56, 0.0, 1.0),   # Violet
    (0.29, 0.0, 0.51),  # Indigo
    (0.0, 0.0, 1.0),    # Blue
    (0.0, 1.0, 0.0),    # Green
    (1.0, 1.0, 0.0),    # Yellow
    (1.0, 0.5, 0.0),    # Orange
    (1.0, 0.0, 0.0),    # Red
]

def init():
    glClearColor(1.0, 1.0, 1.0, 1.0)  # White background
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    gluOrtho2D(0, width, 0, height)

def draw_arc(center_x, center_y, inner_radius, thickness, start_angle, end_angle, color):
    glColor3f(*color)
    glBegin(GL_TRIANGLE_STRIP)
    for angle in range(start_angle, end_angle + 1):
        rad = math.radians(angle)
        x_outer = center_x + (inner_radius + thickness) * math.cos(rad)
        y_outer = center_y + (inner_radius + thickness) * math.sin(rad)
        x_inner = center_x + inner_radius * math.cos(rad)
        y_inner = center_y + inner_radius * math.sin(rad)
        glVertex2f(x_outer, y_outer)
        glVertex2f(x_inner, y_inner)
    glEnd()

def display():
    glClear(GL_COLOR_BUFFER_BIT)
    center_x = width // 2
    center_y = 100
    inner_radius = 100
    arc_thickness = 20
    start_angle = 0
    end_angle = 180

    for i in range(current_band):
        draw_arc(center_x, center_y, inner_radius + i * arc_thickness, arc_thickness, start_angle, end_angle, rainbow_colors[i])

    glFlush()

def update(value):
    global current_band
    if current_band < len(rainbow_colors):
        current_band += 1
        glutPostRedisplay()
        glutTimerFunc(600, update, 0)  # Delay between each band (ms)

def main():
    glutInit()
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
    glutInitWindowSize(width, height)
    glutInitWindowPosition(100, 100)
    glutCreateWindow(b"Animated Rainbow Formation")
    init()
    glutDisplayFunc(display)
    glutTimerFunc(600, update, 0)  # Start animation
    glutMainLoop()

if __name__ == "__main__":
    main()

File: rainbow.cpp
#include <GL/glut.h>
#include <math.h>

void init(void) {
    glClearColor(0.0, 0.0, 0.0, 0.0);
    glMatrixMode(GL_PROJECTION);
    gluOrtho2D(0.0, 400.0, 0.0, 400.0);
}

void drawRainbow(void) {
    glClear(GL_COLOR_BUFFER_BIT);

    float radius = 150.0;
    float centerX = 200.0;
    float centerY = 100.0;
    int segments = 100;

    float colors[7][3] = {
        {1.0, 0.0, 0.0},
        {1.0, 0.5, 0.0},
        {1.0, 1.0, 0.0},
        {0.0, 1.0, 0.0},
        {0.0, 0.0, 1.0},
        {0.3, 0.0, 0.5},
        {0.5, 0.0, 1.0}
    };

    for (int color = 0; color < 7; color++) {
        glColor3fv(colors[color]);
        glLineWidth(10.0);

        glBegin(GL_LINE_STRIP);
        for (int i = 0; i <= segments / 2; i++) {
            float angle = (float)i / (float)segments * 3.14159;
            float x = centerX + radius * cos(angle);
            float y = centerY + radius * sin(angle);
            glVertex2f(x, y);
        }
        glEnd();

        radius -= 15.0;
    }

    glFlush();
}

void display(void) {
    drawRainbow();
}

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(400, 400);
    glutInitWindowPosition(100, 100);
    glutCreateWindow("Rainbow using OpenGL");

    init();
    glutDisplayFunc(display);
    glutMainLoop();

    return 0;
}








// background colors:
//when all are 0.0-black
//when {0.0,1.0,0.0,0.0}-green
//when{ 1.0,0.0,0.0,0.0 }-red
//when {0.0,0.0,1.0,0.0}-blue
//when {0.0,0.0,1.0,1.0}-black
//when {1.0,1.0,1.0,1.0}-white

